sycl/source/detail/program_impl.cpp

```diff
@@ -38,14 +39,34 @@ program_impl::program_impl(
     throw runtime_error("Non-empty vector of programs expected",
                         PI_INVALID_VALUE);
   }
+
+  // Sort the programs to avoid deadlocks due to locking multiple mutexes &
+  // verify that all programs are unique.
+  std::sort(ProgramList.begin(), ProgramList.end(),
+            [](const shared_ptr_class<program_impl> &A,
+               const shared_ptr_class<program_impl> &B) {
+              return A.get() < B.get();
+            });
+  auto It = std::unique(ProgramList.begin(), ProgramList.end(),
+                        [](const shared_ptr_class<program_impl> &A,
+                           const shared_ptr_class<program_impl> &B) {
+                          return A.get() == B.get();
+                        });
+  if (It != ProgramList.end()) {
+    throw runtime_error("Attempting to link a program with itself",
+                        PI_INVALID_PROGRAM);
+  }
+
   MContext = ProgramList[0]->MContext;
   MDevices = ProgramList[0]->MDevices;
   vector_class<device> DevicesSorted;
   if (!is_host()) {
     DevicesSorted = sort_devices_by_cl_device_id(MDevices);
   }
   check_device_feature_support<info::device::is_linker_available>(MDevices);
+  vector_class<unique_ptr_class<std::lock_guard<std::mutex>>> Locks;
   for (const auto &Prg : ProgramList) {
+    Locks.emplace_back(new std::lock_guard<std::mutex>(Prg->MMutex));
```

Comment:
One can eliminate using pointers here via smth like this: `Locks.emplace_back({Prg->MMutex});` with `Locks` as a `list` instead of `vector`.